exp/metric-gen/metric/parser.go

```diff
@@ -0,0 +1,247 @@
+package metric
+
+import (
+	"fmt"
+	"go/ast"
+	"go/types"
+	"sort"
+	"strings"
+
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	"k8s.io/klog/v2"
+	"sigs.k8s.io/controller-tools/pkg/crd"
+	"sigs.k8s.io/controller-tools/pkg/loader"
+	"sigs.k8s.io/controller-tools/pkg/markers"
+
+	"k8s.io/kube-state-metrics/v2/pkg/customresourcestate"
+)
+
+type Parser struct {
+	*crd.Parser
+
+	CustomResourceStates map[schema.GroupKind]customresourcestate.Resource
+	FlattenedMetrics     map[crd.TypeIdent][]customresourcestate.Metric
+}
+
+func NewParser(parser *crd.Parser) *Parser {
+	return &Parser{
+		Parser:               parser,
+		CustomResourceStates: make(map[schema.GroupKind]customresourcestate.Resource),
+		FlattenedMetrics:     make(map[crd.TypeIdent][]customresourcestate.Metric),
+	}
+}
+
+func (p *Parser) NeedResourceFor(groupKind schema.GroupKind) {
```

Comment:
I think we'll need a lot of godoc explaining what all this is doing to make it maintainable for folks who are not super-famiilar with the controller-gen implementation